### ISSUE TITLE
Allow setting the path to the Vala compiler using the VALAC env variable

### ISSUE THREAD

**Issue by astavale** [mesonbuild/meson#1645]
It looks as though the executable for valac (and other compilers) is hard coded in /mesonbuild/environment.py :

```python
def detect_vala_compiler(self):
    exelist = ['valac']
```
When developing the Vala compiler it is useful to build programs against a newly patched version of valac. With C the [HowToX](https://github.com/mesonbuild/meson/wiki/howtox#set-compiler) suggests the `CC` variable can be used:

> Set compiler
> When first running Meson, set it in an environment variable.
> `$ CC=mycc meson <options>`

Is this the best way? So for Vala it would be:
`$ VALAC=/home/development-vala/installed/bin/valac meson <options>`

Another, maybe clearer, approach is to treat this as a [cross compilation](https://github.com/mesonbuild/meson/wiki/Cross-compilation). So a cross file is created with the relevant binaries:

```ini
[binaries]
valac = '/home/development-vala/installed/bin/valac'
```
Then this is compiled with:
`meson build --cross-file development_vala.txt`

Thoughts?

**Comment by nirbheek** [comment]
We should definitely read `VALAC`. I thought we already did that, but I just checked and we don't.

Treating the run as a cross-compilation is wrong though, because it has several non-obvious effects (such as, tests won't be run without an `exe_wrapper`, built programs won't be run, etc). However, we were [discussing a `dependency file`](https://github.com/mesonbuild/meson/issues/1525) of sorts, so maybe it's good to extend that to also include this.

I'll rename this bug to be about `VALAC` specifically. In the meantime, I recommend symlinking that valac inside `~/.local/bin` and adding that to your `PATH`.
